Results_Engine/Query/Group.cs

```diff
@@ -67,6 +70,108 @@ public static partial class Query
+        {
+            if (results == null || !results.Any())
+                return new List<List<T>>();
+
+            return GroupByMultipleSelectors(results, IdentifierFunctions(results.First().GetType(), typeof(ObjectIdentifierAttribute)));
+        }
+
+        /***************************************************/
+
+        [Description("Group results by all identifiers of the object.")]
+        [Input("results", "Results to group.")]
+        [Output("groupedResults", "Results grouped as List of Lists.")]
+        public static List<List<T>> GroupByAllIdentifiers<T>(this IEnumerable<T> results) where T : IResult
+        {
+            if (results == null || !results.Any())
+                return new List<List<T>>();
+
+            return GroupByMultipleSelectors(results, IdentifierFunctions(results.First().GetType(), typeof(IdentifierAttribute)));
+        }
+
+        /***************************************************/
+        /**** Private Methods                           ****/
+        /***************************************************/
+
+        [Description("Groups by multiple selectors by contiously grouping by each provided selector.")]
+        private static List<List<T>> GroupByMultipleSelectors<T>(this IEnumerable<T> results, IEnumerable<Func<object, object>> funcs)
+        {
+            IEnumerable<IEnumerable<T>> listGroup = new List<IEnumerable<T>>() { results };
+            foreach (var func in funcs)
+                listGroup = listGroup.Select(x => x.GroupBy(y => func(y))).SelectMany(x => x);
+
+            return listGroup.Select(x => x.ToList()).ToList();
+        }
+
+        /***************************************************/
+
+        [Description("Gets a list of compiled selectors corresponding to all properties with the attribute type on the type. Allows the attribute type to be defined on an interface of the object type.")]
+        private static List<Func<object, object>> IdentifierFunctions(Type type, Type attributeType)
+        {
+            Tuple<Type, Type> key = new Tuple<Type, Type>(type, attributeType);
+            List<Func<object, object>> funcs;
+            lock (m_funcLock)   //Lock extraction for thread saftey
+            {
+                if (!m_identifierFunctions.TryGetValue(key, out funcs)) //Try get stored pre-compiled functions
+                {
+                    //Get all properties with the provided attribute and compile to functions
+                    funcs = type.PropertiesWithAttribute(attributeType, true, true).Select(x => x.ToFunc()).ToList();
+                    m_identifierFunctions[key] = funcs; //Store functions for future usage
+                }
+            }
+            return funcs;
+        }
+
+        /***************************************************/
+
+        [Description("Compiles the property info into a Func<object,object>.")]
+        private static Func<object, object> ToFunc(this PropertyInfo prop)
```

Comment:
This method looks useful, I would make this a public Reflection_Engine Query method.
If I understood it correctly, a more appropriate naming (especially if we make it a Reflection_Engine public Query method) could be "PropertyGetter" or something similar?

```cs
[Description("Compiles the PropertyInfo into a Func<object,object> corresponding to the property getter.")]
Query.PropertyGetter(this PropertyInfo prop)
```

Reply:
Can move to Base_Engines ToFunc method. COnsidered putting it there, as if fundamentally does something very similar, but in a slightly different way (which was the reason I just put it as a private method in this file to start with). But Think you are right that it can be useful elsewhere. Will move it.